### PR TITLE
reorganize the bzl files to have fewer and cleaner public entry points

### LINKED ISSUE #39: [Long term]: Consolidate into less (public) .bzl files and rules
No action item for now, but we need to start simplifying what user-visible pieces we have in the near future - this is also an issue internally. For example, `dart_ddc_bundle` - maybe that should just be part of a dart_web_application() macro.

Very careless strawman:
- `dart_library`
- `dart_web_application` ... becomes a macro
  - Creates a dart_library for you from srcs/deps
  - Creates a `dart_ddc_bundle` for you
  - Creates a `dart_web_binary` for you
    - Runs dart2js of script_file.dart to script_file.js and script_file.debug.js (?)
    - We will still need flags for configuring dart2js output
  - Creates one or more `dart_dev_server` rules for you
- `dart_vm_binary` (Doesn't change much/at all)

### PR DESCRIPTION
**Pull request by jakemac53**
Partly resolves https://github.com/dart-lang/rules_dart/issues/39.

Maybe I went too far, but to me this organization makes more sense. The public facing bzl files are now just `web.bzl`, `vm.bzl`, `core.bzl`, and `code_gen.bzl`. For all but `code_gen.bzl` I made the public bzl files only contain the basic rule definition, not the implementations. This makes it easier for users to see what is defined in each file and what their attributes are. The implementations of the rules each have their own bzl files, under a new `internal` folder.

Let me know what you guys think :).

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#### 0.2.0
+
+**Breaking Change**: Re-organization of the build rules:
+
+*  `dart_ddc_bundle` moved into `dart/build_rules/web.bzl`.
+*  `dev_server` moved into `dart/build_rules/web.bzl`.
+*  `pub_repositories` and `dart_repositories` have been combined into the
+   `setup_dart_repositories` function, which lives in
+   `dart/build_rules/repositories.bzl`.
+   * `dart/build_rules/pub.bzl` has moved to `dart/build_rules/internal/pub.bzl`
```
Comment (natebosch) [inline CHANGELOG.md]: [nit] indented too far? did you intend for this to be a sublist?

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#### 0.2.0
+
+**Breaking Change**: Re-organization of the build rules:
+
+*  `dart_ddc_bundle` moved into `dart/build_rules/web.bzl`.
+*  `dev_server` moved into `dart/build_rules/web.bzl`.
+*  `pub_repositories` and `dart_repositories` have been combined into the
+   `setup_dart_repositories` function, which lives in
```
Comment (natebosch) [inline CHANGELOG.md]: 'setup_dart_repositories' and 'dart_repositories' seems like a difference without a distinction IMO. Maybe 'dart_repositories' as the public one, then '_pub_repositories' and either '_core_repositories' or '_sdk_repositories'?

**Review by natebosch — APPROVED** [review]
Overall I like the new organization

**Comment by natebosch** [comment]
This will fix https://github.com/dart-lang/rules_dart/issues/50

File: CHANGELOG.md (hunk 2 above)
Comment (jakemac53) [inline CHANGELOG.md]: Ya I think you are right, I will just rename it back to `dart_repositories` and then change the other one to `sdk_repositories`.

File: CHANGELOG.md (hunk 1 above)
Comment (jakemac53) [inline CHANGELOG.md]: I am actually going to just remove this line in favor of just saying `pub_repositories()` no longer exists (even though it does, but thats an internal impl detail)

File: dart/build_rules/web.bzl [hunk 3]
```diff
@@ -14,167 +14,105 @@
...
+# able to serve.
+#
+# If you have renamed the rules_dart_repo, you will need to provide that name
+# via the `rules_dart_repo_name` argument.
+def dev_server(rules_dart_repo_name = "@io_bazel_rules_dart",
+               **kwargs):
+  dart_vm_binary(
+      srcs = [rules_dart_repo_name + "//dart/tools/dev_server:bin/server.dart"],
+      script_file = rules_dart_repo_name + "//dart/tools/dev_server:bin/server.dart",
+      deps = [rules_dart_repo_name + "//dart/tools/dev_server:server"],
```
Comment (natebosch) [inline dart/build_rules/web.bzl]: do we still want `+ deps` here?

**Comment by jakemac53** [comment]
merging for now as this touches so many files its going to make merging with future work difficult, @cbracken @matanlurey feel free to review later

**Pull request merged**